sstables/trie.cc

```diff
@@ -0,0 +1,787 @@
+        _path.push_back({pv_to_reader_node(root_pos, _page), -1});
+        expensive_assert(_path.back().node.pos / cached_file::page_size == _page->pos() / cached_file::page_size);
+    });
+}
+
+// Documented near the declaration.
+template <trie_reader_source Input>
+future<void> trie_cursor<Input>::init(uint64_t root_pos) {
+    check_invariants();
+    reset();
+    return push_page_and_node(root_pos);
+}
+
+// Documented near the declaration.
+template <trie_reader_source Input>
+bool trie_cursor<Input>::try_push(uint64_t pos) {
+    expensive_assert(_path.back().node.pos / cached_file::page_size == _page->pos() / cached_file::page_size);
+    if (_path.back().node.pos / cached_file::page_size == pos / cached_file::page_size) {
+        _path.push_back({pv_to_reader_node(pos, _page), -1});
+        return true;
+    }
+    return false;
+}
+
+// Documented near the declaration.
+template <trie_reader_source Input>
+void trie_cursor<Input>::pop() {
+    _path.pop_back();
+}
+
+
+// Documented near the declaration.
+template <trie_reader_source Input>
+future<follow_result> trie_cursor<Input>::follow(const_bytes key) {
+    check_invariants();
+    expensive_assert(initialized());
+    auto post_check = defer([this] {
+        check_invariants();
+    });
+    expensive_log("follow, root_pos={}, key={}", _path[0].node.pos, fmt_hex(key));
+    expensive_assert(_path.back().child_idx == -1 || eof());
+    // Reset the cursor back to a freshly-initialized, unset state, with only the root in _path.
+    co_await ascend(0);
+    _path.back().child_idx = -1;
+
+    int next_branch_depth = -1;
+    int prev_branch_depth = -1;
+    int prefix_depth = -1;
+    size_t i = 0;
+    while (i < key.size()) {
+        // Fast path for long chains without any payloads or branches.
+        //
+        // const uint8_t* __restrict__ p = reinterpret_cast<const uint8_t* __restrict__>(_page->get_view().data() + _path.back().node.pos % cached_file::page_size);
+        // if (*p == (SINGLE_NOPAYLOAD_4 << 4 | 2) & *(p + 1) == uint8_t(key[i])) {
+        //     const uint8_t* start = p; 
+        //     const uint8_t* beg = p - _path.back().node.pos % cached_file::page_size;
+        //     const size_t keysize = key.size();
+        //     while (p - 32 >= beg && i+16 <= keysize - 1) {
+        //         typedef unsigned char  vector32b  __attribute__((__vector_size__(32)));
+        //         typedef unsigned char  vector16b  __attribute__((__vector_size__(16)));
+        //         vector32b a = {};
+        //         memcpy(&a, p - 32, 32);
+        //         auto z = uint8_t(SINGLE_NOPAYLOAD_4 << 4 | 2);
+        //         vector16b b = {};
+        //         memcpy(&b, &key[i], 16);
+        //         vector16b c = {z, z, z, z, z, z, z, z, z, z, z, z, z, z, z, z};
+        //         vector32b d = __builtin_shufflevector(c, b, 0, 16, 1, 17, 2, 18, 3, 19, 4, 20, 5, 21, 6, 22, 7, 23, 8, 24, 9, 25, 10, 26, 11, 27, 12, 28, 13, 29, 14, 30, 15, 31);
```

Comment:
What's the plan? Predict how the index would look if there were no branches, and see if it happens to match?

Reply:
Yes.